packages/cli/src/chat/chat-service.ts

```diff
@@ -0,0 +1,298 @@
+
+		ws.once('close', async () => {
+			ws.off('message', onMessage);
+			clearInterval(intervalId);
+			this.sessions.delete(key);
+		});
+
+		ws.on('message', onMessage);
+
+		const session: Session = {
+			connection: ws,
+			executionId,
+			sessionId,
+			intervalId,
+			isPublic,
+			lastHeartbeat: Date.now(),
+		};
+
+		this.sessions.set(key, session);
+
+		ws.send('n8n|heartbeat');
+	}
+
+	private outgoingMessageHandler(sessionKey: string) {
+		return async () => {
+			let session: Session | undefined;
+			try {
+				session = this.sessions.get(sessionKey);
+
+				if (!session) return;
+				if (session.isProcessing) {
+					return;
+				}
+				session.isProcessing = true;
+
+				const { connection, executionId, sessionId, waitingNodeName, isPublic } = session;
+
+				if (!executionId || !connection) {
+					session.isProcessing = false;
+					return;
+				}
+
+				const execution = await this.getExecution(executionId, sessionKey);
+
+				if (!execution) {
+					session.isProcessing = false;
+					return;
+				}
+
+				if (waitingNodeName) {
+					const lastNode = getLastNodeExecuted(execution);
+
+					if (execution.status === 'waiting' && lastNode?.name !== waitingNodeName) {
+						connection.send('n8n|continue');
+						session.waitingNodeName = undefined;
+					}
+					session.isProcessing = false;
+					return;
+				}
+
+				if (execution.status === 'waiting') {
+					const message = getMessage(execution);
+
+					if (message !== undefined) {
+						connection.send(message);
+
+						const lastNode = getLastNodeExecuted(execution);
+
+						if (lastNode && shouldResumeImmediately(lastNode)) {
+							connection.send('n8n|continue');
+							const data = { action: 'user', chatInput: '', sessionId };
+							await this.resumeExecution(executionId, data, sessionKey);
+							session.waitingNodeName = undefined;
+						} else {
+							session.waitingNodeName = lastNode?.name;
+						}
+					}
+
+					session.isProcessing = false;
+					return;
+				}
+
+				if (execution.status === 'success') {
+					const shouldNotReturnLastNodeResponse =
+						!isPublic || (isPublic && isResponseNodeMode(execution));
+
+					if (shouldNotReturnLastNodeResponse) {
+						closeConnection(connection);
+						session.isProcessing = false;
+						return;
+					}
+
+					const textMessage = prepareMessageFromLastNode(execution);
+
+					connection.send(textMessage, () => {
+						closeConnection(connection);
+					});
+
+					session.isProcessing = false;
+					return;
+				}
+
+				session.isProcessing = false;
+			} catch (error) {
+				if (session) session.isProcessing = false;
+				this.logger.error(
+					`Error sending message to chat in session ${sessionKey}: ${(error as Error).message}`,
+				);
+			}
+		};
+	}
+
+	private incomingMessageHandler(sessionKey: string) {
+		return async (data: RawData) => {
+			try {
+				const session = this.sessions.get(sessionKey);
+
+				if (!session) return;
+
+				const message = data.toString();
```

Comment:
We are stringifying the data twice. Here and in `processIncomingData`. Can we please unify